Aggregation metrics

sam-data-guy-iam opened the pull request:
Addition of GroupedInstanceMetric and derived metrics @matanor @elronbandel

yoavkatz commented on requirements/base.rqr:
```diff
@@ -8,3 +8,5 @@ scikit-learn
 dpath
 jiwer
 editdistance
+statsmodels
```
Sam.  Are these requirements needed for the change? I don't see their use.

yoavkatz commented on src/unitxt/metrics.py:
```diff
@@ -1430,3 +1431,116 @@ def _compute(
+
+    from statistics import mean
+
+    metric = Accuracy
```
Probably best not to have a default to make sure people override it.

matanor replied:
Also, why not initialize with the class object? Meaning:    `metric = Accuracy()`

yoavkatz commented on src/unitxt/metrics.py:
```diff
@@ -1430,3 +1431,116 @@ def _compute(
+            references, predictions, additional_inputs
+        ):
+            # allow any number of columns to be used as the identifier
+            keyname = tuple(inputs_dict.values())
```
I think the user needs to explicitly  specify which are fields from additional_inputs to take.  There could be many which are irrelevant.

sam-data-guy-iam replied:
In cards, is there a way for the user to specify which fields are relevant to be included in additional_inputs?  I assume perhaps there are cases where additional_inputs fields could be used for calculation of the accuracy metric but NOT the grouping (which the aggregation function operates on).  I assumed that all fields in additional_inputs would be used for grouping.  What did you envision?

yoavkatz commented on src/unitxt/metrics.py:
```diff
@@ -1430,3 +1431,116 @@ def _compute(
+
+    metric = Accuracy
+    group_score_func = mean
+    group_score_name = "mean"
```
Is 'group_score_name' this used?

yoavkatz commented on src/unitxt/metrics.py:
```diff
@@ -1430,3 +1431,116 @@ def _compute(
+    from statistics import mean
+
+    metric = Accuracy
+    group_score_func = mean
```
Maybe call it 'group_score_aggregation_func'?

yoavkatz reviewed (changes requested):
Overall, looks good and useful.   A few requested changes and questions.
I'd prefer @elronbandel will also reviews.

matanor commented on src/unitxt/metrics.py:
```diff
@@ -1430,3 +1431,116 @@ def _compute(
         return result
+
+
+class GroupedInstanceMetric(GlobalMetric):
```
I am not sure that deriving from `GlobalMetric` is suitable here, since the underlying metric is computed per instance. So its a bit confusing with the global metric goal to allow computation which depends on a set of instances, but here we are doing per-instance computations.

I think a better option could be modifying [`InstanceMetric.process(..)`](https://github.com/IBM/unitxt/blob/main/src/unitxt/metrics.py#L373) to support grouping. After the loop that calculates the scores per instance ([this loop](https://github.com/IBM/unitxt/blob/main/src/unitxt/metrics.py#L377)), you could group the scores by the group key, if one is provided. This requires keeping track of the group key per score. Then, the code could loop over the reduction map and do the computation per group. In case there is no group key specified, there will be only one group with all the instances. Then the final step would be to averege over all group scores (which would be one score if no groups are defined). What do you think?

matanor reviewed (changes requested):
hi @sam-data-guy-iam looks good! please see my comments ..

sam-data-guy-iam commented:
After more consideration, I rewrote this as a sub-class of InstanceMetric, as suggested by @matanor, which can now take an InstanceMetric as well as GlobalMetric (and possibly others, not sure), which have a process method.  I added a confidence interval method that resamples the instances and extract the scores, without having to recalculate the instance scores each time as GlobalMetric would have required.  Please take a look and let me know what you think.

yoavkatz commented:
@matanor @elronbandel - Can we proceed with this PR and see if there are any additional needed changes (beside fix to the conflict)

matanor commented:
@sam-data-guy-iam is this PR replaced by your new PR https://github.com/IBM/unitxt/pull/452?

sam-data-guy-iam commented:
Yes

On Tue, Jan 9, 2024, 10:32 AM matanor ***@***.***> wrote:

> @sam-data-guy-iam <https://github.com/sam-data-guy-iam> is this PR
> replaced by your new PR #452 <https://github.com/IBM/unitxt/pull/452>?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/IBM/unitxt/pull/430#issuecomment-1882613114>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5V5WLBVQ2PWDE5BMAUKYLDYNT6B3AVCNFSM6AAAAABBCS6AK6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBSGYYTGMJRGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

matanor commented:
> Yes
> […](#)
> On Tue, Jan 9, 2024, 10:32 AM matanor ***@***.***> wrote: @sam-data-guy-iam <https://github.com/sam-data-guy-iam> is this PR replaced by your new PR #452 <#452>? — Reply to this email directly, view it on GitHub <[#430 (comment)](https://github.com/IBM/unitxt/pull/430#issuecomment-1882613114)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/A5V5WLBVQ2PWDE5BMAUKYLDYNT6B3AVCNFSM6AAAAABBCS6AK6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBSGYYTGMJRGQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

OK, so i am closing this PR..
